framework/aster-frame/src/vm/frame/meta.rs

```diff
@@ -0,0 +1,156 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use core::{
+    ops::Deref,
+    sync::atomic::{AtomicU32, AtomicU8},
+};
+
+use crate::{
+    arch::mm::PagingConsts,
+    vm::{page_size, Paddr, PagingLevel},
+};
+
+pub mod mapping {
+    //! The metadata of each physical frame is mapped to fixed virtual addresses
+    //! from [`FRAME_METADATA_BASE_VADDR`] to [`FRAME_METADATA_CAP_VADDR`]. This
+    //! range is divided into portions in the geometric series of 1/2, which are
+    //! 1/2, 1/4, 1/8, ..., 1/2^N. The Nth portion is used to store the metadata
+    //! of the level N pages.
+
+    use core::mem::size_of;
+
+    use super::FrameMeta;
+    use crate::vm::{
+        kspace::{FRAME_METADATA_BASE_VADDR, FRAME_METADATA_CAP_VADDR},
+        page_size, Paddr, PagingConstsTrait, PagingLevel, Vaddr,
+    };
+
+    /// Convert a physical address of the page to the virtual address of the metadata container.
+    pub const fn page_to_meta<C: PagingConstsTrait>(paddr: Paddr, level: PagingLevel) -> Vaddr {
+        let kvspace = FRAME_METADATA_CAP_VADDR - FRAME_METADATA_BASE_VADDR;
+        let base = FRAME_METADATA_CAP_VADDR - (kvspace >> (level - 1));
+        let offset = paddr / page_size::<C>(level);
+        base + offset * size_of::<FrameMeta>()
+    }
+
+    /// Convert a virtual address of the metadata container to the physical address of the page.
+    pub const fn meta_to_page<C: PagingConstsTrait>(vaddr: Vaddr) -> Paddr {
```

Comment:
This will be also used when doing mapping for the metadata slots, other than getting the physical address from the handle. One of the two usages will suffer from a cast. Taking a `* const FrameMeta` here introduces another cast. So I prefer taking a `Vaddr`.